File: src/python/1_3_fsync.py
from typing import List
import random
import os

def saveData3(path:str, data: List[bytes]):
    temp_file = f"{path}.tmp.{random.randint(1, 10)}"
    
    try:
        with open(temp_file, "wb") as temp:
            temp.write(data)
            os.fsync(temp)
    except Exception:
        os.remove(temp_file)
    
    return os.rename(temp_file, path)
    
if __name__ == "__main__":
    byteArray = bytearray([96, 86])
    saveData3("./test_python", byteArray)

File: src/golang/1_3_fsync.go
package main

import (
	"fmt"
	"math/rand"
	"os"
)

func SaveData3(path string, data []byte) error {
	tmp := fmt.Sprintf("%s.tmp.%d", path, rand.Intn(10))
	fp, err := os.OpenFile(tmp, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0664)
	if err != nil {
		return err
	}
	defer fp.Close()

	_, err = fp.Write(data)
	if err != nil {
		os.Remove(tmp)
		return err
	}

	err = fp.Sync() // fsync
	if err != nil {
		os.Remove(tmp)
		return err
	}

	return os.Rename(tmp, path)
}

func main() {
	byteArray := []byte{96, 86}
	SaveData3("./test", byteArray)
}
